source/opt/local_ssa_elim_pass.cpp

```diff
@@ -0,0 +1,759 @@
+    for (auto ii = bp->begin(); ii != bp->end(); ++ii) {
+      switch (ii->opcode()) {
+        case SpvOpStore: {
+          uint32_t varId;
+          (void) GetPtr(&*ii, &varId);
+          if (!IsTargetVar(varId))
+            break;
+          // Register new stored value for the variable
+          label2ssa_map_[label][varId] =
+              ii->GetSingleWordInOperand(kStoreValIdInIdx);
+        } break;
+        case SpvOpLoad: {
+          uint32_t varId;
+          (void) GetPtr(&*ii, &varId);
+          if (!IsTargetVar(varId))
+            break;
+          uint32_t replId = 0;
+          const auto ssaItr = label2ssa_map_.find(label);
+          if (ssaItr != label2ssa_map_.end()) {
+            const auto valItr = ssaItr->second.find(varId);
+            if (valItr != ssaItr->second.end())
+              replId = valItr->second;
+          }
+          // If variable is not defined, use undef
+          if (replId == 0) {
+            replId = Type2Undef(GetPointeeTypeId(def_use_mgr_->GetDef(varId)));
+          }
+          // Replace load's id with the last stored value id for variable
+          // and delete load. Kill any names or decorates using id before
+          // replacing to prevent incorrect replacement in those instructions.
+          const uint32_t loadId = ii->result_id();
+          KillNamesAndDecorates(loadId);
+          (void)def_use_mgr_->ReplaceAllUsesWith(loadId, replId);
+          def_use_mgr_->KillInst(&*ii);
+          modified = true;
+        } break;
+        default: {
+        } break;
+      }
+    }
+    visitedBlocks_.insert(label);
+    // Look for successor backedge and patch phis in loop header
+    // if found.
+    uint32_t header = 0;
+    bp->ForEachSuccessorLabel([&header,this](uint32_t succ) {
+      if (visitedBlocks_.find(succ) == visitedBlocks_.end()) return;
+      assert(header == 0);
+      header = succ;
+    });
+    if (header != 0)
+      PatchPhis(header, label);
+  }
+  // Remove all target variable stores.
+  for (auto bi = func->begin(); bi != func->end(); ++bi) {
+    for (auto ii = bi->begin(); ii != bi->end(); ++ii) {
+      if (ii->opcode() != SpvOpStore)
+        continue;
+      uint32_t varId;
+      (void) GetPtr(&*ii, &varId);
+      if (!IsTargetVar(varId))
+        break;
+      assert(!HasLoads(varId));
+      DCEInst(&*ii);
+      modified = true;
+    }
+  }
+  return modified;
+}
+
+void LocalSSAElimPass::Initialize(ir::Module* module) {
+
+  module_ = module;
+
+  // Initialize function and block maps
+  id2function_.clear();
+  id2block_.clear();
+  block2structured_succs_.clear();
+  for (auto& fn : *module_) {
+    id2function_[fn.result_id()] = &fn;
+    for (auto& blk : fn)
+      id2block_[blk.id()] = &blk;
+  }
+
+  // Initialize Target Type Caches
+  seen_target_vars_.clear();
+  seen_non_target_vars_.clear();
+
+  // Initialize set of variables only referenced by supported operations
+  supported_ref_vars_.clear();
+
```

Comment:
I was presuming that a pass was only Process'ed once.

Would all the pass-level collection names be preceeded by _., as in _.visitedBlocks_?

It doesn't seem like there is a lot of difference between enumerating the collections in the Context or in Initialize.

For the moment I will clear them all in Initialize() and we can refine from there.